NitroxPatcher/Patches/Dynamic/Constructable_Construct_Patch.cs

```diff
@@ -1,62 +1,185 @@
             {
-                lastTargetBase = null;
-                lastTargetBaseOffset = default(Int3);
+                CoroutineHost.StartCoroutine(postSpawner);
             }
+        }
+        // update as a normal module
 
-            return true;
+        // TODO: A latest packet is always sent even after the base is fully built, we don't want that
+        // When #2021 is merged, just use ThrottledPacketSender.RemovePendingPackets
+        Resolve<ThrottledPacketSender>().SendThrottled(new ModifyConstructedAmount(entityId, amount),
+            (packet) => { return packet.GhostId; }, 0.1f);
+    }
+
+    public static IEnumerator BroadcastObjectBuilt(ConstructableBase constructableBase, NitroxId entityId)
+    {
+        BaseGhost baseGhost = constructableBase.model.GetComponent<BaseGhost>();
+        constructableBase.SetState(true, true);
+        if (!baseGhost.targetBase)
+        {
+            Log.Error("Something wrong happened, couldn't find base after finishing building ghost");
+            yield break;
+        }
+        Base targetBase = baseGhost.targetBase;
+        Base parentBase = null;
+        if (constructableBase.tr.parent)
+        {
+            parentBase = constructableBase.GetComponentInParent<Base>(true);
         }
 
-        public static void Postfix(Constructable __instance, bool __result)
+        // If a module was spawned we need to transfer the ghost id to it for further recognition
+        BuildUtils.TryTransferIdFromGhostToModule(baseGhost, entityId, constructableBase, out GameObject moduleObject);
+
+        // Have a delay for baseGhost to be actually destroyed
+        yield return null;
+
+        if (parentBase)
         {
-            if (!__result) return;
+            // update existing base
+            if (!parentBase.TryGetNitroxId(out NitroxId parentId))
+            {
+                // TODO: Probably add a resync here
+                Log.Error("Parent base doesn't have a NitroxEntity, which is not normal");
+                yield break;
+            }
+            MoonpoolManager moonpoolManager = parentBase.gameObject.EnsureComponent<MoonpoolManager>();
 
-            NitroxServiceLocator.LocateService<Building>().ChangeConstructionAmount(__instance.gameObject, __instance.constructedAmount);
-            if (__instance._constructed)
+            GlobalRootEntity builtPiece = null;
+            if (moduleObject)
             {
-                NitroxServiceLocator.LocateService<Building>().ConstructionComplete(__instance.gameObject, Optional.OfNullable(lastTargetBase), lastTargetBaseOffset, lastFace);
+                if (moduleObject.TryGetComponent(out IBaseModule builtModule))
+                {
+                    builtPiece = InteriorPieceEntitySpawner.From(builtModule);
+                }
+                else if (moduleObject.GetComponent<VehicleDockingBay>())
+                {
+                    builtPiece = moonpoolManager.LatestRegisteredMoonpool;
```

Comment:
Shouldn't builtPiece be checked for null here bc it's possible to be never set?

Reply:
This is quite a general code piece right here. Not only modules but also normal base pieces go through this code so it's fine to have builtPiece null